src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +192,153 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
   }
 )
 
-export const selectRewardsAmountByDenom = createSelector(
+export const selectAllRewardsByAssetId = createDeepEqualOutputSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAssetId,
+  (stakingData, selectedAssetId): Record<string, chainAdapters.cosmos.Reward[]> => {
+    if (!stakingData || !stakingData.rewards) return {}
+
+    const rewards = stakingData.rewards.reduce(
+      (acc: Record<string, chainAdapters.cosmos.Reward[]>, current: ValidatorReward) => {
+        if (!acc[current.validator.address]) {
+          acc[current.validator.address] = []
+        }
+
+        acc[current.validator.address].push(
+          ...current.rewards.filter(x => x.assetId === selectedAssetId)
+        )
+
+        return acc
+      },
+      {}
+    )
+
+    return rewards
+  }
+)
+
+export const selectRewardsAmountByAssetId = createSelector(
   selectRewardsByAccountSpecifier,
-  selectDenom,
-  (rewardsByAccountSpecifier, denom): string => {
+  selectAssetId,
+  (rewardsByAccountSpecifier, selectedAssetId): string => {
     if (!rewardsByAccountSpecifier.length) return ''
 
-    const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
+    const rewards = rewardsByAccountSpecifier.find(rewards => rewards.assetId === selectedAssetId)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byValidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byValidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const getUndelegationsAmountByValidatorAddress = memoize(
+  (
+    allUndelegationsEntries: Record<string, chainAdapters.cosmos.UndelegationEntry[]>,
+    validatorAddress: string
+  ) => {
+    return get(
+      allUndelegationsEntries,
+      validatorAddress,
+      [] as chainAdapters.cosmos.UndelegationEntry[]
```

Comment:
do not cast here, `get` is generic https://stackoverflow.com/questions/57824356/how-to-use-generic-types-with-lodash-get

Reply:
[] is of type `never[]`, so we're forced to fill generics to `get`, this is so verbose for such a little thing, as you prefer anyway!

Reply:
> [] is of type `never[]`, so we're forced to fill generics to `get`

Could you elaborate on this @NeOMakinG ? Generics are first-class citizens of lodash typings in `DefinitelyTyped`, and it is possible to type `get<TObject, TKey>()`:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/object.d.ts#L1032

The first type argument is `TObject` i.e the type of the object, the second is `TKey` i.e the type of the key (as a keyof TObject), so this should work (untested, you might also be able to leave out the second type argument):

```javascript
get<Record<string, chainAdapters.cosmos.UndelegationEntry[]>, string>(allUndelegationsEntries, validatorAddress, [])
```

Regarding it being "so verbose for such a little thing", we should keep in mind that generics are really type-level, whereas **castings are never a drop-in replacement for typings**.  They do have some valid use cases (type guards, tests where you don't want to clutter the module with the full shape of a JSON response...) but you're basically escaping/lying to the compiler when using them, which is usually a sign of wrong/missing typings except for said valid use cases.

Reply:
@gomesalexandre It's already fixed as def1 asked, I used generics everywhere